File: 0697-degree-of-an-array/0697-degree-of-an-array.py
class Solution(object):
    def minOf(self, a, b):
        if a <= b: 
            return a
        return b

    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        count = {}
        firstSeen = {}

        n = len(nums)
        maxCount = 0
        length = 1

        for i in range(n):
            if nums[i] not in firstSeen:
                firstSeen[nums[i]] = i + 1

            count[nums[i]] = count.get(nums[i], 0) + 1

            currentCount = count[nums[i]]
            firstSeenCount = firstSeen[nums[i]]

            if currentCount > maxCount:
                maxCount = currentCount
                length = (i + 1) - firstSeenCount + 1
            elif currentCount == maxCount:
                length = self.minOf(length, (i + 1) - firstSeenCount + 1)

        return length

File: 0697-degree-of-an-array/0697-degree-of-an-array.cpp
class Solution {
public:
    int minOf(int a, int b) {
        if(a <= b) return a;
        return b;
    }
    int findShortestSubArray(vector<int>& nums) {
        unordered_map<int,int> count;
        unordered_map<int,int> firstSeen;

        int maxCount =0; int n = nums.size();
        int length = 1;

        for(int i=0; i<n; i++) {
            if(!firstSeen[nums[i]]) firstSeen[nums[i]] = i+1;

            count[nums[i]]++ ;

            if(count[nums[i]] > maxCount) {
                maxCount = count[nums[i]];
                length = (i+1) - firstSeen[nums[i]] + 1;
            }
            else if(count[nums[i]] == maxCount) {
                length = minOf(length, (i+1) - firstSeen[nums[i]] +1);
            }
        }
        return length;
    }
};